src/decoder.rs

```diff
@@ -545,117 +544,145 @@ pub enum ScopeEntry {
     Decoder(Decoder),
 }
 
-pub struct Scope<'a> {
-    parent: Option<&'a Scope<'a>>,
-    names: Vec<Label>,
-    entries: Vec<ScopeEntry>,
+pub enum Scope<'a> {
+    Empty,
+    Multi(&'a MultiScope<'a>),
+    Single(SingleScope<'a>),
+    Decoder(DecoderScope<'a>),
 }
 
-pub struct ScopeIter<'a> {
-    parent: Option<&'a Scope<'a>>,
-    name_iter: std::iter::Rev<std::slice::Iter<'a, Label>>,
-    entry_iter: std::iter::Rev<std::slice::Iter<'a, ScopeEntry>>,
+pub struct MultiScope<'a> {
+    parent: &'a Scope<'a>,
+    entries: Vec<(Label, Value)>,
 }
 
-impl<'a> Iterator for ScopeIter<'a> {
-    type Item = (&'a Label, &'a ScopeEntry);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        match (self.name_iter.next(), self.entry_iter.next()) {
-            (Some(name), Some(entry)) => Some((name, entry)),
-            _ => match self.parent {
-                Some(parent) => {
-                    *self = parent.into_iter();
-                    self.next()
-                }
-                None => None,
-            },
-        }
-    }
+pub struct SingleScope<'a> {
+    parent: &'a Scope<'a>,
+    name: &'a str,
+    value: &'a Value,
 }
 
-impl<'a> IntoIterator for &'a Scope<'a> {
-    type Item = (&'a Label, &'a ScopeEntry);
+pub struct DecoderScope<'a> {
+    parent: &'a Scope<'a>,
+    name: &'a str,
+    decoder: Decoder,
+}
 
-    type IntoIter = ScopeIter<'a>;
+impl<'a> Scope<'a> {
+    fn get_value_by_name(&self, name: &str) -> &Value {
+        match self {
+            Scope::Empty => panic!("value not found: {name}"),
+            Scope::Multi(normal) => normal.get_value_by_name(name),
+            Scope::Single(single) => single.get_value_by_name(name),
+            Scope::Decoder(decoder) => decoder.parent.get_value_by_name(name),
+        }
+    }
 
-    fn into_iter(self) -> Self::IntoIter {
-        ScopeIter {
-            parent: self.parent,
-            name_iter: self.names.iter().rev(),
-            entry_iter: self.entries.iter().rev(),
+    fn get_decoder_by_name(&self, name: &str) -> &Decoder {
+        match self {
+            Scope::Empty => panic!("decoder not found: {name}"),
+            Scope::Multi(normal) => normal.parent.get_decoder_by_name(name),
+            Scope::Single(single) => single.parent.get_decoder_by_name(name),
+            Scope::Decoder(decoder) => decoder.get_decoder_by_name(name),
         }
     }
-}
 
-impl<'a> Scope<'a> {
-    pub fn new() -> Self {
-        let parent = None;
-        let names = Vec::new();
-        let entries = Vec::new();
-        Scope {
-            parent,
-            names,
-            entries,
+    pub fn get_bindings(&self, bindings: &mut Vec<(Label, ScopeEntry)>) {
+        match self {
+            Scope::Empty => {}
+            Scope::Multi(normal) => normal.get_bindings(bindings),
```

Comment:
Though purely cosmetic, the hold-over binding identifier `normal` is a bit misplaced here...

Reply:
Fixed.